File: c#/19-Remove-Nth-Node-from-End-of-List.cs
#region Iteration (Two Pass)
// Time O(n)
// Space O(1)
public class Solution
{
    public ListNode RemoveNthFromEnd(ListNode head, int n)
    {
        // Dummy node simplifies edge cases
        var dummy = new ListNode(0, head);
        int len = 0;
        var cur = head;

        // Step 1: Find length
        while (cur != null)
        {
            cur = cur.next;
            len++;
        }

        // Step 2: Stop at node before nth from end
        cur = dummy;
        for (int i = 0; i < len - n; i++)
        {
            cur = cur.next;
        }

        // Step 3: Remove node and return head
        cur.next = cur.next.next;
        return dummy.next;
    }
}
#endregion

File: python/19-Remove-Nth-Node-From-End-of-List.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head

        left, right = dummy, dummy

        for _ in range(n + 1):
            right = right.next

        while right is not None:
            right = right.next
            left = left.next

        left.next = left.next.next

        return dummy.next
